### PR TITLE
add: document scoping, document preapproval, evidence validation

### PR DESCRIPTION
**Pull request by mickhansen**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -54,6 +58,78 @@ mutation {
...
+
+When using preapproval document-scoping is automatically used so make sure that you pass all documents from the signature order to the signatory.
+
+```
+mutation {
+  addSignatory(
+    input: {
+      signatureOrderId: "[signatureOrder.id]"
+      documents: [
+        {id: "[signatureOrder.documents[0].id]", preapproved: true}
```
Comment (sgryt) [inline README.md:94]: How would this work when one wants to include all the documents (so not hardcoding the indices)

**Comment by sgryt** [comment]
If I set up the signatory like this

image

I end up being able to open the document but the Approve button is then a no-op.

**Comment by sgryt** [comment]
If I try to sign with a mismatching evidence claim value, I get

image

Would be surprising if the user could change anything by trying again here, it is not likely that they have more than 1 CPR (or similar) to try with.

File: README.md (hunk 1 above)
Comment (mickhansen) [reply README.md:94]: @sgryt You would simply omit the `documents` argument. Or if preapproving all documents you'd need a map/loop in the client code.

**Comment by mickhansen** [comment]
> If I set up the signatory like this
> 
> image
> 
> I end up being able to open the document but the Approve button is then a no-op.
> 
> Perhaps we should either make it impossible to open the document(s), or change the approve/deny buttons to just a Close button ?

You're not seeing a active sign button right from the beginning?

The frontend changes are not quite in yet. If all documents are preapproved then the user will be redirected directly to Criipto authentication.

**Comment by mickhansen** [comment]
> If I try to sign with a mismatching evidence claim value, I get
> 
> image
> 
> Would be surprising if the user could change anything by trying again here, it is not likely that they have more than 1 CPR (or similar) to try with.

We'll be adding a custom error screen for this case that better explains the issue.

This is specifically to add cases like Insurance for couples where the couple might share an email address - And the husband might open the wife's link and use his own NemID, so he could actually retry by then using his wifes NemID. (We'll also have some text explaining that he probably clicked the wrong link, Trine is working on the design)

**Comment by mickhansen** [comment]
Can hold off on merging this untill the frontend PR is ready as we can use the netlify preview just fine.

**Comment by sgryt** [comment]
> > If I set up the signatory like this
> > image
> > I end up being able to open the document but the Approve button is then a no-op.
> > Perhaps we should either make it impossible to open the document(s), or change the approve/deny buttons to just a Close button ?
> 
> You're not seeing a active sign button right from the beginning?
> 
> The frontend changes are not quite in yet. If all documents are preapproved then the user will be redirected directly to Criipto authentication.

The Sign button is there, but if I open the document anyway, closing it again seems to be harder than it could be.

**Comment by mickhansen** [comment]
> The Sign button is there, but if I open the document anyway, closing it again seems to be harder than it could be.

Makes sense to swap it with a Close button if the document is preapproved.

**Comment by sgryt** [comment]
> > If I try to sign with a mismatching evidence claim value, I get
> > image
> > Would be surprising if the user could change anything by trying again here, it is not likely that they have more than 1 CPR (or similar) to try with.
> 
> We'll be adding a custom error screen for this case that better explains the issue.
> 
> This is specifically to add cases like Insurance for couples where the couple might share an email address - And the husband might open the wife's link and use his own NemID, so he could actually retry by then using his wifes NemID. (We'll also have some text explaining that he probably clicked the wrong link, Trine is working on the design)

Ah, OK, I see. Had not thought about that possibility, but certainly a relevant use case.

**Comment by sgryt** [comment]
> Can hold off on merging this untill the frontend PR is ready as we can use the netlify preview just fine.

As long as the task is recorded somewhere so we don't forget it, I don't think we need to hold back the merge of the demo code.

File: README.md (hunk 1 above)
Comment (sgryt) [reply README.md:94]: > @sgryt You would simply omit the `documents` argument. Or if preapproving all documents you'd need a map/loop in the client code.

Oh, I did not know that the `graphql` construct supported loops. Good to know :)

**Comment by mickhansen** [comment]
> > > If I try to sign with a mismatching evidence claim value, I get
> > > image
> > > Would be surprising if the user could change anything by trying again here, it is not likely that they have more than 1 CPR (or similar) to try with.
> > 
> > 
> > We'll be adding a custom error screen for this case that better explains the issue.
> > This is specifically to add cases like Insurance for couples where the couple might share an email address - And the husband might open the wife's link and use his own NemID, so he could actually retry by then using his wifes NemID. (We'll also have some text explaining that he probably clicked the wrong link, Trine is working on the design)
> 
> Ah, OK, I see. Had not thought about that possibility, but certainly a relevant use case.

That is the specific use-case that GF presented, atleast :)

**Comment by mickhansen** [comment]
> > Can hold off on merging this untill the frontend PR is ready as we can use the netlify preview just fine.
> 
> As long as the task is recorded somewhere so we don't forget it, I don't think we need to hold back the merge of the demo code.

It's the next thing on my list in general :)

File: README.md (hunk 1 above)
Comment (mickhansen) [reply README.md:94]: @sgryt Well the query itself does not support loops, but you can use an input variable like `documents: $documents` and your code can then use a loop to generate that variable.

**Pull request merged**